src/bindings/node/node.cc

```diff
@@ -0,0 +1,181 @@
+    napi_status status;
+    napi_property_descriptor properties[] = {
+        DECLARE_NAPI_METHOD("route", Route),
+    };
+
+    napi_value cons;
+    status = napi_define_class(env, "Actor", NAPI_AUTO_LENGTH, New, nullptr, 1, properties, &cons);
+    assert(status == napi_ok);
+
+    status = napi_create_reference(env, cons, 1, &constructor);
+    assert(status == napi_ok);
+
+    status = napi_set_named_property(env, exports, "Actor", cons);
+    assert(status == napi_ok);
+    return exports;
+  }
+  static void Destructor(napi_env env, void* nativeObject, void* finalize_hint) {
+    delete reinterpret_cast<Actor*>(nativeObject);
+  }
+
+private:
+  explicit Actor(const char* config) : actor(make_conf(config)), env_(nullptr), wrapper_(nullptr) {
+  }
+  ~Actor() {
+    napi_delete_reference(env_, wrapper_);
+  }
+
+  static napi_value New(napi_env env, napi_callback_info info) {
+    napi_status status;
+
+    napi_value target;
+    status = napi_get_new_target(env, info, &target);
+    assert(status == napi_ok);
+    bool is_constructor = target != nullptr;
+
+    if (is_constructor) {
+      // Invoked as constructor: `new Actor(...)`
+      size_t argc = 1;
+      napi_value jsthis;
+      napi_value argv[1];
+
+      status = napi_get_cb_info(env, info, &argc, argv, &jsthis, NULL);
+      checkNapiStatus(status, env, "Failed to parse arguments");
+
+      // parse config string passed in
+      size_t buffer_size;
+      status = napi_get_value_string_utf8(env, argv[0], nullptr, 0, &buffer_size);
+      checkNapiStatus(status, env, "Failed to get config string length");
+
+      if (buffer_size > 1024 * 1024) {
+        napi_throw_error(env, NULL, "Too large JSON config");
+      }
+
+      char buffer[++buffer_size];
+      status = napi_get_value_string_utf8(env, argv[0], buffer, buffer_size, &buffer_size);
+      checkNapiStatus(status, env, "Failed to get config string");
+
+      Actor* obj = new Actor(buffer);
+
+      obj->env_ = env;
+      status = napi_wrap(env, jsthis, reinterpret_cast<void*>(obj), Actor::Destructor,
+                         nullptr, // finalize_hint
+                         &obj->wrapper_);
+      assert(status == napi_ok);
+
+      return jsthis;
+    } else {
+      // Invoked as plain function `Actor(...)`, turn into construct call.
+      size_t argc_ = 1;
+      napi_value args[1];
+      status = napi_get_cb_info(env, info, &argc_, args, nullptr, nullptr);
+      assert(status == napi_ok);
+
+      const size_t argc = 1;
+      napi_value argv[argc] = {args[0]};
+
+      napi_value cons;
+      status = napi_get_reference_value(env, constructor, &cons);
+      assert(status == napi_ok);
+
+      napi_value instance;
+      status = napi_new_instance(env, cons, 0, NULL, &instance);
+      assert(status == napi_ok);
+
+      return instance;
+    }
+  }
+
+  static napi_value Route(napi_env env, napi_callback_info info) {
+    napi_status status;
```

Comment:
can we take all the boilerplate here that lets you get `reqString` and move that into another private method. then just call that at the beginning of this function like:

   auto reqString = ParseRequest(env, info);

and then at the bottom another boilerplate wrapper to turn an `std::string` into a js string:

    auto myStr = WrapString(route_json, env);

this slight refactor will allow adding all the other methods really quickly since they are all the exact same pattern.

Reply:
oh yeah, I did start doing that and got distracted, will do 👍